packages/webpack/lib/middleware/render.js

```diff
@@ -1,77 +1,40 @@
 'use strict';
 
 const { join } = require('path');
-const EventEmitter = require('events');
 
 const webpack = require('webpack');
 const MemoryFS = require('memory-fs');
 
 const sourceMapSupport = require('source-map-support');
 const requireFromString = require('require-from-string');
 
+const { Resolvable } = require('../utils/resolvable');
+
 module.exports = exports = function createRenderMiddleware(webpackConfig) {
-  const transpiler = exports.createTranspiler(webpackConfig);
-  let middleware, error;
-  transpiler.on('start', () => {
-    middleware = null;
-    error = null;
-  });
-  transpiler.on('success', (result) => {
-    middleware = result;
-    error = null;
+  const resolvable = new Resolvable();
+  const compiler = Object.assign(webpack(webpackConfig), {
+    outputFileSystem: new MemoryFS(),
   });
-  transpiler.on('error', (result) => {
-    middleware = null;
-    error = result;
-  });
-  const getMiddlewarePromise = () => {
-    if (error) {
-      return Promise.reject(error);
-    }
-    if (middleware) {
-      return Promise.resolve(middleware);
-    }
-    return new Promise((resolve) => {
-      transpiler.once('result', () => {
-        resolve(getMiddlewarePromise());
-      });
-    });
-  };
-  return function webpackRenderMiddleware(req, res, next) {
-    getMiddlewarePromise()
-      .then((middleware) => middleware(req, res, next))
-      .catch(next);
-  };
-};
-
-exports.createTranspiler = function createTranspiler(webpackConfig) {
-  const emitter = new EventEmitter();
-  const compiler = webpack(webpackConfig);
-  compiler.outputFileSystem = new MemoryFS();
-  compiler.hooks.watchRun.tap('untool-transpiler', () => emitter.emit('start'));
-  const emitResult = (...args) => {
-    emitter.emit(...args);
-    emitter.emit('result');
-  };
   const handleCompilation = (compileError, stats) => {
     if (compileError) {
-      emitResult('error', compileError);
+      resolvable.reject(compileError);
     } else if (stats.hasErrors()) {
-      stats.toJson().errors.forEach(emitResult.bind(null, 'error'));
+      stats.toJson().errors.forEach(resolvable.reject);
```

Comment:
With regards to multiple errors you are completely right. Since this is code that only ever runs in developemnt or during static builds I considered that acceptable. If we wanted to change that, we would have to build some middleware capable of handling multiple errors...

Your recollection is spot on: whatever error is being passed to `resolvable.reject()` here is being passed to Express' `next()` later on.